File: Vesuv.Core/core/logger.cpp
#include "logger.h"

namespace Vesuv::Core {

	using namespace System;
	using namespace System::Diagnostics;


	Logger::Logger(String^& context) :
		context(context) {}


	Logger::~Logger() {}


	Logger^ Logger::GetFor() {
		StackTrace stackTrace(1, true);
		int i = 0;
		StackFrame^ stackFrame = stackTrace.GetFrame(i);
		String^ context = stackFrame->GetMethod()->DeclaringType->FullName;
		return GetFor(context);
	}
	Logger^ Logger::GetFor(String^ context) {
		return gcnew Logger(context);
	}


	void Logger::Log() {
		Debug::WriteLine("");
	}
	void Logger::Log(String^ message) {
		Debug::WriteLine(message, context);
	}

}

File: Vesuv.Core/Core/Logger.cs
using System.Diagnostics;

namespace Vesuv.Core
{

	// Todo: Implement a better Logging Solution
	public class Logger :
		CoreObject
	{

		private readonly string context;

		public Logger(string context) {
			this.context = context;
		}

		protected override void Dispose(bool disposing) {
			if (disposing) {
			}
			base.Dispose(disposing);
		}

		public static Logger GetFor() {
			var stackTrace = new StackTrace(1, true);
			var context = stackTrace.GetFrame(0).GetMethod().DeclaringType.FullName;
			return GetFor(context);
		}
		public static Logger GetFor(string context) {
			return new Logger(context);
		}


		public virtual void Log() {
			Debug.WriteLine("");
		}
		public virtual void Log(string message) {
			Debug.WriteLine(message, this.context);
		}

		public virtual void Warn() {
			Debug.WriteLine("");
		}
		public virtual void Warn(string message) {
			Debug.WriteLine(message, this.context);
		}

		public virtual void Error() {
			Debug.WriteLine("");
		}
		public virtual void Error(string message) {
			Debug.WriteLine(message, this.context);
		}
	}

}
